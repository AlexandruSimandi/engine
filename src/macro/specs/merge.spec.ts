import pluginTester from 'babel-plugin-tester';
import prettier from 'prettier';
import plugin from 'babel-plugin-macros';

const macroFile = "'./build/macro/index.macro'";

pluginTester({
  plugin,
  babelOptions: { filename: __filename },
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: 'babel'
    });
  },
  tests: {
    'should support Merge': {
      code: `
        import { producer } from ${macroFile}
        producer(({
          a1 = Merge.foo,
          a2 = Merge['@foo'],
          a3 = Merge['$a2'],
          a4 = Merge.foo['@bar.baz'],
          a5 = Merge.foo.bar['$a2'],
          a6 = {
            baz: Merge.foo['$a4'].baz,
          },
          a7 = Merge.foo['$a5.baz'],
          a8 = Merge.foo[':prop'],
          a9 = Merge.foo[':prop'].baz,
          a10 = Merge.foo[':prop'].baz['@a2']
        }) => { })
      `,
      snapshot: true
    }
  }
});
