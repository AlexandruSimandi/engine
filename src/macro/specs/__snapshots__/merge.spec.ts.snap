// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros should support Merge: should support Merge 1`] = `

import { producer } from './build/macro/index.macro'
producer(({
  a1 = Merge.foo,
  a2 = Merge['@foo'],
  a3 = Merge['$a2'],
  a4 = Merge.foo['@bar.baz'],
  a5 = Merge.foo.bar['$a2'],
  a6 = {
    baz: Merge.foo['$a4'].baz,
  },
  a7 = Merge.foo['$a5.baz'],
  a8 = Merge.foo[':prop'],
  a9 = Merge.foo[':prop'].baz,
  a10 = Merge.foo[':prop'].baz['@a2']
}) => { })

      ↓ ↓ ↓ ↓ ↓ ↓

producer({
  args: {
    type: "STRUCT",
    value: {
      a1: {
        type: "MERGE",
        path: [
          {
            type: "CONST",
            value: "foo"
          }
        ]
      },
      a2: {
        type: "MERGE",
        path: [
          {
            type: "EXTERNAL",
            path: ["foo"]
          }
        ]
      },
      a3: {
        type: "MERGE",
        path: [
          {
            type: "INTERNAL",
            path: ["a2"]
          }
        ]
      },
      a4: {
        type: "MERGE",
        path: [
          {
            type: "CONST",
            value: "foo"
          },
          {
            type: "EXTERNAL",
            path: ["bar", "baz"]
          }
        ]
      },
      a5: {
        type: "MERGE",
        path: [
          {
            type: "CONST",
            value: "foo"
          },
          {
            type: "CONST",
            value: "bar"
          },
          {
            type: "INTERNAL",
            path: ["a2"]
          }
        ]
      },
      a6: {
        type: "STRUCT",
        value: {
          baz: {
            type: "MERGE",
            path: [
              {
                type: "CONST",
                value: "foo"
              },
              {
                type: "INTERNAL",
                path: ["a4"]
              },
              {
                type: "CONST",
                value: "baz"
              }
            ]
          }
        }
      },
      a7: {
        type: "MERGE",
        path: [
          {
            type: "CONST",
            value: "foo"
          },
          {
            type: "INTERNAL",
            path: ["a5", "baz"]
          }
        ]
      },
      a8: {
        type: "MERGE",
        path: [
          {
            type: "CONST",
            value: "foo"
          },
          {
            type: "INVOKE",
            path: "prop"
          }
        ]
      },
      a9: {
        type: "MERGE",
        path: [
          {
            type: "CONST",
            value: "foo"
          },
          {
            type: "INVOKE",
            path: "prop"
          },
          {
            type: "CONST",
            value: "baz"
          }
        ]
      },
      a10: {
        type: "MERGE",
        path: [
          {
            type: "CONST",
            value: "foo"
          },
          {
            type: "INVOKE",
            path: "prop"
          },
          {
            type: "CONST",
            value: "baz"
          },
          {
            type: "EXTERNAL",
            path: ["a2"]
          }
        ]
      }
    }
  },
  fn: ({ a1, a2, a3, a4, a5, a6, a7, a8, a9, a10 }) => {}
});


`;
